### DISCUSSION TITLE
Compare dynamic field values of given index queries using elastalert. Is it possible?

### DISCUSSION THREAD

**Discussion by machillef** [jertel/elastalert2#598] [Q&A] [answered]
Hello all.

Lately I have been testing elastalert(along with Praeco as a web GUI) as a tool for automating alerts using Elasticsearch as a central log server.
However, some of the alerts I would like to perform are more complex than simple string queries.
Let me give you an example.
I have an index called test-index with Source/Destination IP pairs as the fields(among other irrelevant fields):
```
Doc1:
sourceIP: ipA
destIP: ipE

Doc2:
sourceIP: ipB
destIP: ipD

Doc3:
sourceIP: ipC
destIP: ipE
```
I would like to run a complex query, for example find all results where the destination IP is the same, and if the total count is more than 10 send an alert. In my example index, the query would return Doc1 and Doc3, but would not fire an alert.
As you can guess, since the IP(both source and destination) are dynamic I can't use something like "if destIP = "192.168.0.1" or something similar. It would have to be generic.

Is this something that I could do using Elastalert or because dives into more complex queries I can not?

In any case, thank you very much for all the work you have done.

**Comment by jertel** [comment, marked as answer]
Hello. Yes, you can use ElastAlert 2 for this. Review the [frequency](https://elastalert2.readthedocs.io/en/latest/ruletypes.html?highlight=cardinality#frequency) documentation, combined with the `query_key` parameter to alert when some count of the same dest_ip is detected within a specified timeframe.

**Reply by machillef** [reply]
Thank you very very much, this was exactly what I was looking for.

**Reply by machillef** [reply]
Hello again,

Happy new Year! I wanted to ask another question, which is kinda like a step up from my initial one, so I am not sure if I should open a new thread or not. If you want me I can ask it in a separate one.
So, given the above configuration, what I would like to do now is to compare 2 fields and have an alert when the total count is more than 10.
For example: If sourceIP is the same and destIP is the same and total count > 10 send an alert.
What I did was create a query_key with 2 keys, both source and destination IP, and I created a simple script to send 25 logs(json formatted) to test it out:
15 of them have sourceIP: ipA and destIP: ipE
5 of them have sourceIP: ipB and destIP: ipE
5 of them have sourceIP: ipA and destIP: ipC

However when I test it out I get a notification for all of them, where I would expect to only get an alert from the first pair of 15logs, and not the rest.
The relevant code I use is the following:
```
type: "frequency"
num_events: 10 
query_key:
  - "ipsource.keyword"
  - "ipdestination.keyword"
  ```
  Did I misunderstood the usage of query_key by any chance? 
  
  Thank you in advance!

**Reply by nsano-rururu** [reply]
- Why don't you look at the documentation?
- Why not provide the entire rule configuration information?
- Why don't you look at the project guidelines?

**Reply by nsano-rururu** [reply]
@machillef
https://elastalert2.readthedocs.io/en/latest/recipes/faq.html#how-can-i-prevent-duplicate-alerts

**Reply by machillef** [reply]
Hello,

Thank you very much for your reply. I read the documentation you sent me as well as the original query_key documentation, and my understanding is that when using a compound of multiple fields(i.e when creating the new field field1,field2 = value1,value2), you need the num_events to match both of these, correct? So in the following code
```
name: "Test source and dest rule"
num_events: 10
priority: 2
query_key:
  - "ipsource.keyword"
  - "ipdestination.keyword"
realert:
  minutes: 5
smtp_port: 25
terms_size: 50
timeframe:
  minutes: 5
timestamp_field: "@timestamp"
timestamp_type: "iso"
type: "frequency"
use_strftime_index: false
```
It should only match if there are more than 10 instances of field1,field2 == value1,value2 right?
In my example I have the following
15 num_events of ipsource.keyword,ipdestination.keyword == ipA,ipE
5   num_events of ipsource.keyword,ipdestination.keyword == ipB,ipE
5  num_events of ipsource.keyword,ipdestination.keyword == ipA,ipC

Maybe the calculation is done based on the value themselves? So, because the value ipA appears more than 10 times, I get an alert for both pairs? Same as ipE value
Apologies if I completely misunderstood the documentation.

**Reply by nsano-rururu** [reply]
@machillef
https://elastalert2.readthedocs.io/en/latest/recipes/faq.html#how-can-i-prevent-duplicate-alerts

**Reply by nsano-rururu** [reply]
Why don't you look at the documentation?

**Reply by nsano-rururu** [reply]
I don't use elastalert2, so I will answer with what I read from the document.

If the values of ipsource.keyword and ipdestination.keyword are the same combination and there are more than 10 matches in 5 minutes, an alert is sent.
Since the realert is set, even if the same match occurs in 5 minutes
Since it is treated as a duplicate alert, the alert is not sent.

**Reply by nsano-rururu** [reply]
I have added the function of elastalert2, but I haven't really used it.

**Reply by nsano-rururu** [reply]
I don't have any conditions with filter. Forgot to write?

**Reply by nsano-rururu** [reply]
15 num_events of ipsource.keyword,ipdestination.keyword == ipA,ipE

**Reply by nsano-rururu** [reply]
query_key is called in sql
GROUP BY
    ipsource.keyword
   , ipdestination.keyword

**Reply by nsano-rururu** [reply]
There is a setting called use_terms_query, but there seems to be a bug
https://github.com/Yelp/elastalert/issues/3026

You'll need to check the official documentation to see if a fix is needed.

**Reply by machillef** [reply]
Hello. Thanks for the replies. Even though I haven't checked "use_terms_query" I will give it a look even if there is a bug.
So far, my main issue is that I receive more alerts than I intend to. With my current understanding and the example I have posted I should be only notified once given the following input data:
15 num_events of ipsource.keyword,ipdestination.keyword == ipA,ipE
5 num_events of ipsource.keyword,ipdestination.keyword == ipB,ipE
5 num_events of ipsource.keyword,ipdestination.keyword == ipA,ipC

but I get an alert for all 3 pairs of ipsource,ipdestination. If it was true, I wouldn't mind, so for me its not a re-alert issue, but rather I can not seem to understand why given the above rule and this input data, all 3 pairs of ipsource,ipdestination return True.
maybe it has something to do with SQL, since I am not very familiar with it.

Thank you for all the trouble you are going through!

**Reply by nsano-rururu** [reply]
I'm not familiar with elasticsearch, so I just replaced it with SQL and explained it. If you don't, forget it.

**Reply by nsano-rururu** [reply]
It's midnight in Japan time, so I go to bed. I don't have time to be alone with you. You should do as much as specify `es_debug_trace` to see what kind of query is being issued.
https://elastalert2.readthedocs.io/en/latest/recipes/faq.html?highlight=es_debug_trace%20#my-rule-is-not-getting-any-hits

**Reply by nsano-rururu** [reply]
use_terms_query: If set to true, it seems to use Elasticsearch'aggregation query', but to use this setting,'doc_type' and'query_key' settings are required. I think you have already read it because it is written in the document.

use_terms_query: When true is not specified, it seems that it is counting in the program.

**Reply by nsano-rururu** [reply]
I don't know why you say 3 pairs. Two are specified in query_key. You have specified two compound keys, right? .. It doesn't make sense.

**Reply by machillef** [reply]
Hello. By pairs I meant the sample data:
15 num_events of ipsource.keyword,ipdestination.keyword == ipA,ipE
5 num_events of ipsource.keyword,ipdestination.keyword == ipB,ipE
5 num_events of ipsource.keyword,ipdestination.keyword == ipA,ipC

You are indeed correct, I use query_key with 2 fields, ipsource, and ipdestination. 
Basically I am trying to understand why the query return an alert for each pair of ipsource,ipdestination even when the total num_events is lower than 10.
I will soon try the debugging procedure you send me.
Thank you very much.

**Reply by nsano-rururu** [reply]
As I pointed out earlier, is it correct that the filter is not specified? ..

**Reply by nsano-rururu** [reply]
It is possible that a huge amount of data is being generated, and instead of being treated as duplicate alerts, it is treated as an alert generation exceeding 10 for multiple value pairs. I'm guessing because I haven't seen the data ...

**Reply by nsano-rururu** [reply]
Also check the elastalert_status index.
https://elastalert2.readthedocs.io/en/latest/recipes/faq.html#how-can-i-prevent-duplicate-alerts
>The data for when an alert will fire again is stored in Elasticsearch in the `elastalert_status` index, with a _type\ ``ofsilence`` and also cached in memory.

**Reply by nsano-rururu** [reply]
https://github.com/Yelp/elastalert/issues/720#issuecomment-248723822

**Reply by jertel** [reply]
It's possible that the handling of `query_keys` differs between frequency rule alerting and realerting. Notice in the docs for `frequency` rules, it states:
> query_key: Counts of documents will be stored independently for each value of query_key. Only num_events documents, all with the same value of query_key, will trigger an alert.

Which correlates with what @machillef is observing.

**Reply by nsano-rururu** [reply]
https://github.com/Yelp/elastalert/pull/143

**Reply by nsano-rururu** [reply]
As with other tests, base_test.py doesn't have 0,1,2 cases for alert.call_count, so I think it's better to add it later.

3 query_keys
https://github.com/jertel/elastalert2/blob/master/tests/base_test.py#L632
>ea.rules[0]['query_key'] = 'this,that,those'

2 query_keys
https://github.com/jertel/elastalert2/blob/master/tests/loaders_test.py#L380
>assert test_rule_copy['query_key'] == 'field1,field2'

**Reply by jertel** [reply]
After re-reading that merged PR I agree with you @nsano-rururu; based on what is described it does appear that the compound query keys should be working for frequency alerting. So perhaps there's a bug.

**Reply by nsano-rururu** [reply]
@jertel 

There are multiple query_keys, but it seems that the following yelp / elastalert issue is also wrong. .. AAdd multi_terms aggs](https://github.com/elastic/elasticsearch/pull/67597)  was found in Elasticsearch pull request. It seems to be supported since Elasticsearch 7.12, so it seems that you need to look at the documentation added by Add multi_terms aggs and fix it.

The problem is that I don't know what to do with Elasticsearch 7.12 or earlier, and what to do with Elasticsearch 7.12 or earlier is the fix as pointed out in the issue of yelp / elastalert.

>There is a setting called use_terms_query, but there seems to be a bug
>https://github.com/Yelp/elastalert/issues/3026

**Reply by machillef** [reply]
Hello again. So after troubleshooting on my side, it seems one of the issues was the type of ipsouce and ipdestination.
In Elasticsearch these values were saved as text, and for some reason, when using query_key and observing the elastalert_status index I noticed that the new compound field match_body.ipsource,ipdestination had a value of None,None.
I changed the types to float and so far it seems to be working. I will keep on testing and hope for the best.
Thanks.
